Added ability to specify target in `meson compile`

TheQwertiest opened the pull request:
Implements and fixes #6740

Target name has the following syntax: `[PATH/]FOO[:TARGET_TYPE]`:
- `compile FOO` - works if there is only a single instance of FOO in introspection, errors out otherwise.
- `compile FOO:TARGET_TYPE` - works if there is only a single instance of FOO in introspection with TARGET_TYPE.
- `compile PATH/FOO:TARGET_TYPE` - always works

PATH - relative path from meson root to the target (e.g `./foo` or `subdir/subdir/foo`).
FOO - target name from meson syntax (e.g. `shared_library('foo', ...)`)
TARGET_TYPE - target type (`shared_library`, `static_library`, `jar`, `run` and etc).

Supported backends: msvs and ninja.
Limitations: 
- msvs: `run_target` is not supported

TheQwertiest commented:
CI failures do not seem relevant: all failed configurations failed on the "Update MSYS" step

dcbaker commented on mesonbuild/mcompile.py:
```diff
@@ -27,9 +29,45 @@
 
+def generate_ninja_target_name(target_name: str, builddir: Path, introspect_data: dict) -> str:
+    for target in introspect_data:
+        if (not target['subproject'] and target['name'] == target_name):
```
please don't wrap if statements in parens unless they span multiple lines, this is python not C :)

dcbaker commented on mesonbuild/mcompile.py:
```diff
@@ -27,9 +29,45 @@
+                if (target['filename'] == 'no_name'):
+                    raise MesonException('Can\'t invoke target `{}`: unexpected target configuration: missing filename'.format(target_name))
+                 
+                return str(Path(target['filename'][0]).relative_to(builddir.resolve()))
```
shouldn't this be in the form:
```python
elif target['filename'] == 'no_name':
     raise ...
else:
     return ...
```

TheQwertiest replied:
Well, I prefer making a clear distinction between the purpose of each if\else group.
E.g. I would've used `elif` here if it was handling the same condition parameter like this:
```
if target.name == 'a':
  ...
elif target.name == 'b':
  ...
```
I can change it if you want though, since it's just a style choice.

dcbaker replied:
I guess either way then

dcbaker commented on mesonbuild/mcompile.py:
```diff
@@ -83,14 +120,22 @@ def run(options: 'argparse.Namespace') -> int:
         cmd = [runner, '-C', bdir.as_posix()]
 
+        if options.targets and options.clean:
+            raise MesonException('`target` and `clean` can\'t be used simultaneously')
```
argparse has builtin functionality for this: https://docs.python.org/3/library/argparse.html#mutual-exclusion

TheQwertiest replied:
Thanks for the tip! Fixed

TheQwertiest replied:
This results in `error: argument TARGET: not allowed with argument --clean`, when invoking like this: `meson compile -C dir --clean`.
It seems that `TARGET` is always defined as present even it's just an empty default value.

dcbaker replied:
hmmmm, I'm not sure off the top of my head. It may not be possible to make a switch argument and positional argument mutally exclusive. You may need to either drop my suggestion or use something like `-t/--target=foo`.

I'm fine with whichever way you want to go. Though if decide to not use the mutually exclusive arguments, can we move the argument exclusive checking into common code so each backend function doesn't need to implement it?

TheQwertiest replied:
Removed `mutual_exclusion` and extracted the check to the common section.

dcbaker commented on mesonbuild/mcompile.py:
```diff
@@ -83,14 +120,22 @@ def run(options: 'argparse.Namespace') -> int:
+        if options.targets and options.clean:
+            raise MesonException('`target` and `clean` can\'t be used simultaneously')
+        if options.targets:
+            with (bdir/'meson-info'/'intro-targets.json').open() as f:
```
`/` is an operator, and we always use spaces around operators.

TheQwertiest replied:
Fixed

dcbaker commented on mesonbuild/mcompile.py:
```diff
@@ -83,14 +120,22 @@ def run(options: 'argparse.Namespace') -> int:
+        if options.targets:
+            with (bdir/'meson-info'/'intro-targets.json').open() as f:
+                schema = json.load(f)
+            for target in options.targets:
+                cmd.extend([generate_ninja_target_name(target, bdir, schema)])
```
```suggestion
            cmd.extend([generate_ninja_target_name(t, bdir, schema) for t in options.targets])
```

TheQwertiest replied:
Fixed

dcbaker commented on mesonbuild/mcompile.py:
```diff
@@ -105,11 +150,19 @@ def run(options: 'argparse.Namespace') -> int:
+        if options.targets:
+            with Path(bdir/'meson-info'/'intro-targets.json').open() as f:
+                schema = json.load(f)
+            for target in options.targets:
+                cmd.extend(['/t:{}'.format(generate_msvs_target_name(target, bdir, schema))])
```
```suggestion
            cmd.extend(['/t:{}'.format(generate_msvs_target_name(t, bdir, schema)) for t in options.targets])
```

TheQwertiest replied:
Fixed

TheQwertiest commented:
Thanks for the review! As you might've guessed I'm not very experienced in Python ^^"...

TheQwertiest commented:
@dcbaker , what do you think about replacing backend divination code with introspection data parsing? This should make backend detection more robust.

TheQwertiest commented:
@dcbaker , seems I'll have to rework this PR a bit due to the new info from @jpakkane (https://github.com/mesonbuild/meson/issues/6740#issuecomment-632719896)

TheQwertiest commented:
Reworked PR a bit: 
- Added handling of non-unique names by adding the following syntax to target name: https://github.com/mesonbuild/meson/issues/6740#issuecomment-632725990
- Fixed target execution in subdirs on VS backend.
- Converted parsed introspection data to dict from array to avoid needless data iterations.

TheQwertiest commented:
How do I fix the typing error? 
https://github.com/mesonbuild/meson/pull/7181/checks?check_run_id=700324080#step:5:8
If I add typing after the variable name like this: `var: typing = ...`, it will fail on another agent with lower Python version :\

TheQwertiest commented:
Test failures seems to be unrelated

TheQwertiest commented:
@dcbaker , Thansk!
PS: Force pushes are for test-fixing only (trying to divine output library names...), i.e. no changes to mcompile.
PPS: Should I squish/clean the PR?

TheQwertiest commented:
Rebased to remove conflicts

jpakkane commented:
Needs a release note snippet and the documentation needs to be updated, at least the way of defining build target names.

TheQwertiest commented:
> Needs a release note snippet

Will be done.

> documentation needs to be updated

There is no documentation for `meson compile` at all.
I was planning on making a separate `Meson commands` page in a separate PR, that lists all commands with short description and links to the respective pages (if any).

TheQwertiest commented:
Added a release snippet.
The proper documentation page for all commands will be added in a separate PR (if you allow me to =)).

TheQwertiest commented:
Wait! I've just noticed that target types from introspection have different names from targets in `meson.build` (e.g. `custom_target` in meson.build vs `custom` in introspection). Should I use those instead? If so, than should there be a mapping for meson.build target types such as `library`, `both_libraries` or `build_target`?

TheQwertiest commented:
@jpakkane , https://github.com/mesonbuild/meson/pull/7217 contains documentation for `meson compile` and this PR (among other things).

nioncode commented:
> 
> Limitations:
>    *  ninja: custom_target is not supported
>    *  msvs: run_target is not supported
> 

@TheQwertiest why is this the case, i.e. what's the problem with supporting those? It would be nice to have full support for all targets.

TheQwertiest commented:
@nioncode , I agree, but it's not possible (yet?):
See the following comments:
https://github.com/mesonbuild/meson/blob/547e29718b67d6b1fc22d42e846fee8bc106f475/mesonbuild/mcompile.py#L185
https://github.com/mesonbuild/meson/blob/547e29718b67d6b1fc22d42e846fee8bc106f475/mesonbuild/mcompile.py#L176

nioncode commented:
Regarding msbuild: I'm fairly certain that you can simply build the `vcxproj` directly with `msbuild` as I have done in my old PR: https://github.com/nioncode/meson/blob/8008cc97ee8a7a02b393dc62e7513d6be6b9ffc8/mesonbuild/mcompile.py#L58

Regarding ninja: isn't this what this PR is there for and should thus provide a way to build a custom_target? At least in my cmake projects I have various custom_target/run_target like commands that I execute with `cmake --build` all the time. Is there anything really preventing us from mapping a target name to the correct ninja commands to invoke? I haven't touched meson code for over a year, so I might be completely wrong, but I think it should be possible to know the commands to run for each target, otherwise the build.ninja file could not be generated.

TheQwertiest commented:
> I'm fairly certain that you can simply build the vcxproj directly with msbuild

Yes, but that won't build project dependencies (AFAIK). It will build only the specified project,

> Regarding ninja: isn't this what this PR is there for and should thus provide a way to build a custom_target? 

This PR is for all targets not just custom_target. Custom target is just the odd man out.

Regarding custom target: I've checked the resulting ninja file and there is nothing for the custom target itself. Only for the files it produces, So to make it work one has to gather all custom_target's output and convert them to ninja targets.

nioncode commented:
For my use case (which always involves building on Windows with msbuild + Linux with ninja), this basically means that mcompile can not be used for either `run_target` or `custom_target` (because either of those will fail on one of my platforms), which is unfortunate. I think we should go the extra mile here and gather the required dependencies for vs and invoke them in addition to the `run_target` (or you sure that the dependencies are not being run then? I really can't remember from my PR, but I think those properly work with the vcxproj that we generate in the vs backend) and doing something similar for ninja and `custom_target`.

Does mcompile support `alias_target`s? If this works, then this might be good enough, since we can suggest to wrap `run_target` and `custom_target` in an `alias_target` then.

TheQwertiest commented:
Regarding msvs, I stand corrected: I've just retested and it seems to work even when targeting individual projects. I swear it was not working before! >_<
Maybe something there was a bug in vs backend deps formation somewhere, i.e. it was not a vs problem but rather a problem with vcxproj generation.

TheQwertiest commented:
There is another problem though: manual vcxproj invocation requires a separate msbuild run. I.e. you can't invoke `shared_library` target and `run_target` via a single msbuild invocation.

TheQwertiest commented:
Added support for custom_targets in ninja backend: it was actually quite simple. My bad.

nioncode commented:
Looks great.

> There is another problem though: manual vcxproj invocation requires a separate msbuild run. I.e. you can't invoke shared_library target and run_target via a single msbuild invocation. It will also make all run_target builds sequential. And this will require a total rewrite of mcompile.

Couldn't this be solved by running all `run_target`s in sequence before doing the 'final' msbuild call that builds everything else (if there is a target remaining)? This shouldn't be too hard, since it would basically be:
```python
for t in run_targets:
  run_msbuild(t)
run_msbuild(remaining_targets)
```

Maybe it is also possible to list multiple vcxproj on the command line (this would allow to collapse the msbuild invocations into a maximum of 2 calls), but I'm not sure.

TheQwertiest commented:
> Maybe it is also possible to list multiple vcxproj on the command line (this would allow to collapse the msbuild invocations into a maximum of 2 calls), but I'm not sure.

This's not possible =(
They even have a specific error message for that (smth like "only one project can be specified").

> Couldn't this be solved by running all run_targets in sequence before doing the 'final' msbuild call that builds everything else (if there is a target remaining)?

It will still require:
- a total separation of vs and ninja invocation logic.
- a total rewrite of vs logic.

And it creates another potential problem (if/when https://github.com/mesonbuild/meson/pull/7183 is merged): user can no longer deduce what's actually happening inside, so backend args that he passes might not be applicable anymore if he specifies even a single run_target (since it might be passed not to a single `msbuild .sln` invocation, but to multiple `msbuild .vcxproj` invocations as well).

PS: Yes, these are all excuses for me not to do extra work =)

nioncode commented:
Maybe it is good enough for now to simply say: if a `run_target` is specified (for vs only), no other targets are allowed? Then at least it is possible to run each target one after another.

I don't really understand your points about the backend_args PR though. If I supply backend_args, I'd assume that they apply to every msbuild invocation that is being done. If I don't want to build everything with the same settings, I'd do multpile `meson compile` invocations.

jpakkane commented:
> Maybe it is good enough for now to simply say: if a run_target is specified (for vs only), no other targets are allowed? Then at least it is possible to run each target one after another.

Fine by me.

TheQwertiest commented:
@jpakkane , all done. The only question remaining is `TARGET_TYPE` naming: currently intro target types are used, which are not the same as meson target types (differences in **bold**):

current (based on introspection):
- 'executable',
- 'static_library',
- 'shared_library',
- 'shared_module',
- 'custom',
- 'run',
- 'jar',

meson.build:
- 'executable',
- 'static_library',
- 'shared_library',
- 'shared_module',
- **'custom_target',**
- **'run_target',**
- 'jar',
- **'library'**,
- **'both_libraries'**,
- **'build_target'**,

TheQwertiest commented:
One solution would be just to list all intro-style types in docs.

dcbaker commented on docs/markdown/Commands.md:
```diff
@@ -132,32 +132,50 @@ meson configure builddir -Doption=new_value
+positional arguments:
+  TARGET                                Targets to build. Target has the
+                                        following format: [PATH_TO_TARGET/]TARGE
+                                        T_NAME[:TARGET_TYPE]
```
You should document that target is new in 0.55.0

TheQwertiest replied:
Fixed

dcbaker commented on mesonbuild/mcompile.py:
```diff
@@ -40,61 +43,184 @@ def get_backend_from_coredata(builddir: Path) -> str:
     return coredata.load(str(builddir)).get_builtin_option('backend')
 
-def get_parsed_args_ninja(options: 'argparse.Namespace', builddir: Path):
+def parse_introspect_data(builddir: Path) -> dict:
```
Please make the returned type match the type of `parsed_data`

TheQwertiest replied:
Fixed

dcbaker commented on mesonbuild/mcompile.py:
```diff
@@ -40,61 +43,184 @@ def get_backend_from_coredata(builddir: Path) -> str:
+            'custom',
+            'run',
+            'jar',
+        ]
```
you should use a set rather than a list here, since order doesn't matter and you're just doing membership tests.

TheQwertiest replied:
Fixed

TheQwertiest commented:
@dcbaker , what do you think about target type naming discrepancy?

dcbaker commented:
You mean the meson may generate conflicting target names and your solution of `NAME[:TYPE]`?

TheQwertiest commented:
I mean https://github.com/mesonbuild/meson/pull/7181#issuecomment-644069430 =)
I.e. (intro target type name) != (meson.build target type name)

dcbaker commented:
The only concern I have about that is that introspect is listing what the build_target *is* (shared_library, static_library, etc), but meson.build is listing what it *could be*. that means that the values passed to meson compile could change based on your configuration, which seems less than ideal. ie,
```console
meson setup build -Ddefault_library=static
meson compile build librarytarget:shared_library
```
could fail to compile because `librarytarget` is a `static_library`

I'm not really concerned about the `run` vs `run_target` differences.

TheQwertiest commented:
> The only concern I have about that is that introspect is listing what the build_target is

Right, and it doesn't look like I can check if it was specified as `both_libraries` in `meson.build` (not through introspection files anyway).

> I'm not really concerned about the run vs run_target differences.

Which was actually my main concern :D

TheQwertiest commented:
Rebased and squashed (since it's a huge PITA to rebase otherwise)

TheQwertiest commented:
@dcbaker , is there anything else I need to fix to make it mergeable?

jpakkane commented:
⚔️ conflicts.

TheQwertiest commented:
Rebased

TheQwertiest commented:
PS: Not sure how to fix CI failure - it seems that it can't find `meson` binary on Macos and Windows agents. Is there a way to require python packages from tests?

jpakkane commented:
Let's merge this with the caveat that if it turns out in actual usage that there is a major design flaw, we can do a breaking fix in the next release.
